### PR TITLE
Share shm and mqueue between containers.

### LINKED ISSUE #12046: shared IPC namespace does not share /dev/shm
If I create container A, then create container B with --ipc=container:A, the two containers do not share a /dev/shm mount.  I think this will prevent POSIX shared memory from working.

@mrunalp
@vishh

### PR DESCRIPTION
**Pull request by calavera**
These changes were reverted because they caused inconsistencies in our CI.
This PR brings the changes back so we can fix the underlaying issue.

See #12159
Fixes #12046

/cc @mrunalp

### CODE REVIEW

**Comment by calavera** [comment]
This is finally ready to merge again. The issue that caused CI to fail was a race between containers recovered after the daemon crash and the cleanup logic for ipc mounts. It has been fixed already and all tests pass.

File: daemon/container.go [hunk 1]
```diff
@@ -296,10 +296,17 @@ func (container *Container) Start() (err error) {
 		return err
 	}
 
+	if !(container.hostConfig.IpcMode.IsContainer() || container.hostConfig.IpcMode.IsHost()) {
```
Comment (tiborvass) [inline daemon/container.go]: Would be easier for me to read `!X && !Y` than `!(X || Y)` but it's my personal taste, feel free to disagree

File: daemon/container.go [hunk 2]
```diff
@@ -358,6 +365,10 @@ func (container *Container) isNetworkAllocated() bool {
 func (container *Container) cleanup() {
 	container.releaseNetwork()
 
+	if err := container.unmountIpcMounts(); err != nil {
+		logrus.Errorf("%v: Failed to umount ipc filesystems: %v", container.ID, err)
```
Comment (tiborvass) [inline daemon/container.go]: `%s` for ID

File: daemon/container.go (hunk 1 above)
Comment (calavera) [inline daemon/container.go]: I'll change it, but I didn't write that code :wink:

**Comment by calavera** [comment]
@tiborvass comments addressed, but of course jenkins is very sad.

File: integration-cli/docker_cli_run_test.go [hunk 3]
```diff
@@ -2007,6 +2007,11 @@ func (s *DockerSuite) TestRunModeIpcContainer(c *check.C) {
 	if parentContainerIpc != out {
 		c.Fatalf("IPC different with --ipc=container:%s %s != %s\n", id, parentContainerIpc, out)
 	}
+
+	catOutput, _ := dockerCmd(c, "run", fmt.Sprintf("--ipc=container:%s", id), "busybox", "cat", "/dev/shm/test")
```
Comment (tiborvass) [inline integration-cli/docker_cli_run_test.go]: only if you need to update the PR: `"--ipc=container:"+id` no need for Sprintf

**Comment by tiborvass** [comment]
@calavera I think it'd be great to have a cli test for `mqueue`

**Comment by calavera** [comment]
@tiborvass I'm not completely sure how to test that mqueue works as expected without compiling a C program that sends messages to the queue and checking the queue in the other container, is that what you're asking for? It sounds out of the scope of this PR.

**Comment by tiborvass** [comment]
@calavera ok no worries
weak LGTM (would like more maintainers to approve)

**Comment by mrunalp** [comment]
@calavera @tiborvass Yeah, checking mqueue won't be easy.

**Comment by calavera** [comment]
> weak LGTM (would like more maintainers to approve)

@tiborvass just a reminder that this feature was already approved and merged into master. I only made some changes to make sure the cleanup process works as expected, in a3e6333. I don't think we should go back into the debate of whether we want this or not, but we should check that my changes to the original PR are good.

**Comment by tiborvass** [comment]
@calavera ok thanks. will review that commit then

**Comment by estesp** [comment]
@calavera overall this looks reasonable, specifically your improvements/cleanup.  I just realized there are some issues with the original patch that will have to be dealt with in #12648 since the `os.MkdirAll()` use will make unreadable directories when user namespaces are in use.. but there is no use holding this up as the functionality needed to make that easy is not merged (the `idtools` package) as it is part of #12648.  I can rebase after this is merged and add those changes.

**Comment by calavera** [comment]
This is green again. Please, take a look.

**Comment by estesp** [comment]
LGTM

**Comment by jfrazelle** [comment]
LGTM

**Pull request merged**